cpp/cow_sim/main1.cpp

```diff
@@ -176,20 +219,114 @@ class TensorImpl {
   Storage storage_;
 };
 
+namespace torch {
+// Clone/View defined to refer to them, they are not modified
+TensorImpl clone(TensorImpl self) { return TensorImpl(); }
+TensorImpl view(TensorImpl self) { return TensorImpl(self.storage());}
+
+enum ReshapeArgs { View, Copy };
+// Simulating changes to the impl of an op w/ conditional view behavior
+TensorImpl reshape(TensorImpl self, ReshapeArgs args) {
+  if (args == ReshapeArgs::View) {
+    // this would be added inside the branch where we are creating a view e.g
+    // https://github.com/pytorch/pytorch/blob/12a69afa6d4c4e9720d1b7644291eadb49792e8a/aten/src/ATen/native/TensorShape.cpp#L1670
+    self.storage().maybe_enable_cow_sim_();
+    return view(self);
+  } else {
+    return clone(self);
+  }
+}
+}
+
+// for easy to read examples
+auto tensor() { return TensorImpl(); }
+using torch::view;
+using torch::reshape;
+auto mutates_input(TensorImpl& t) { return t.mutable_data(); }
+auto reads_from_input(TensorImpl& t) {return t.const_data();}
+using torch::ReshapeArgs;
+
+auto example_1() {
+  auto a = tensor();
+  auto b = reshape(a, ReshapeArgs::View);
+  mutates_input(a);
+  reads_from_input(b);
+}
+
+auto example_2() {
+  auto a = tensor();
+  auto c = tensor();
+  c.storage() = a.storage();
+  auto b = reshape(a, ReshapeArgs::View);
+  mutates_input(b);
+  reads_from_input(c);
+}
+
+auto example_3() {
+    TensorImpl a;
+    TensorImpl b = view(a);
+    mutates_input(a);
+    mutates_input(b);
+    reads_from_input(a);
+    reads_from_input(b);
+}
+
+auto example_4() {
+  auto a = tensor();
+  auto b = view(a);
+  auto c = reshape(b, ReshapeArgs::View);
+  mutates_input(b);
+  reads_from_input(a);
+}
+
+auto example_5() {
+  auto a = tensor();
+  auto b = reshape(a, ReshapeArgs::View);
+  auto c = view(b);
+  mutates_input(b);
+  reads_from_input(a);
```

Comment:
Just heads up, as written, this should cause the warning, but it is marked "0" below

Reply:
thanks I was going pretty quick.